storage/src/commit_log/mod.rs

```diff
@@ -0,0 +1,430 @@
+
+        Ok(Location(out.0, out.1))
+    }
+
+    fn get(&self, location: &Location) -> Result<S::Value, CommitLogError> {
+        let cl = self
+            .cl_handle(S::name())
+            .ok_or(CommitLogError::MissingCommitLog { name: S::name() })?;
+        let cl = cl.read().expect("Read lock failed");
+        let bytes = cl.read(location.0, location.1)?;
+        let value = S::Value::decode(&bytes)?;
+        Ok(value)
+    }
+}
+
+pub fn fold_consecutive_locations(locations: &[Location]) -> Vec<Range> {
+    if locations.is_empty() {
+        Vec::with_capacity(0)
+    } else {
+        let mut ranges = vec![];
+
+        let mut prev = locations[0];
+        let mut range = Range(prev.0, prev.1, 1);
+        for curr in &locations[1..] {
+            if curr.is_consecutive(&prev) {
+                range.1 += curr.1;
+                range.2 += 1;
+            } else {
+                ranges.push(range);
+                range = Range(curr.0, curr.1, 1);
+            }
+            prev = *curr;
+        }
+        ranges.push(range);
+
+        ranges
+    }
+}
+
+/// Provides access to all registered commit logs via a log family reference.
+pub struct CommitLogs {
+    base_path: PathBuf,
+    commit_log_map: RwLock<HashMap<String, CommitLogRef>>,
+}
+
+impl CommitLogs {
+    pub(crate) fn new<P, I>(path: P, cfs: I) -> Result<Self, CommitLogError>
+    where
+        P: AsRef<Path>,
+        I: IntoIterator<Item = CommitLogDescriptor>,
+    {
+        let myself = Self {
+            base_path: path.as_ref().into(),
+            commit_log_map: RwLock::new(HashMap::new()),
+        };
+
+        for descriptor in cfs.into_iter() {
+            Self::register(&myself, descriptor.name())?;
+        }
+
+        Ok(myself)
+    }
+
+    /// Register a new commit log.
+    fn register(&self, name: &str) -> Result<(), CommitLogError> {
+        let path = self.base_path.join(name);
+        if !Path::new(&path).exists() {
+            std::fs::create_dir_all(&path)?;
+        }
+        let log = CommitLog::new(path, true)?;
+
+        let mut commit_log_map = self.commit_log_map.write().unwrap();
+        commit_log_map.insert(name.into(), Arc::new(RwLock::new(log)));
+
+        Ok(())
+    }
+
+    /// Retrieve handle to a registered commit log.
+    #[inline]
+    fn cl_handle(&self, name: &str) -> Option<CommitLogRef> {
+        let commit_log_map = self.commit_log_map.read().unwrap();
+        commit_log_map.get(name).cloned()
+    }
+
+    /// Flush all registered commit logs.
+    pub fn flush(&self) -> Result<(), CommitLogError> {
+        let commit_log_map = self.commit_log_map.read().unwrap();
+        for commit_log in commit_log_map.values() {
+            let mut commit_log = commit_log.write().unwrap();
+            commit_log.flush()?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for CommitLogs {
+    fn drop(&mut self) {
+        let _ = self.flush().expect("Failed to flush commit logs");
```

Comment:
@mambisi 
no expect, expect in drop is kind of undefined behavior
better:
```
let _ = self.flush();
```

or 

```
if let Err(e) = self.flush() {
    println!("Failed to flush commit logs, reason: {:?}", e);
}
```